theano/tensor/nnet/nnet.py

```diff
@@ -999,7 +999,7 @@ def c_code_template(dtype):
         {
             if (NULL != %(nll)s) Py_XDECREF(%(nll)s);
             %(nll)s = (PyArrayObject*)PyArray_SimpleNew(1,
-                PyArray_DIMS(%(y_idx)s), PyArray_TYPE((PyArrayObject*) py_%(x)s));
+                PyArray_DIMS(%(y_idx)s), PyArray_TYPE(%(y_idx)s));
```

Comment:
the make_node() method tell that the nll output have the same dtype as x, not as y_idx. So just change y_idx to x in the part that you changed. Good catch.